pkg/gvkcompare/gvkcompare.go

```diff
@@ -0,0 +1,276 @@
+	err = r.collectNamespaceReport(unsupportedGVRs)
+	if err != nil {
+		return errors.Wrap(err, "Unable to evaluate GVR gaps for migrated resources")
+	}
+
+	return nil
+}
+
+func (r *GVKCompare) PrepareSourceDiscovery(c client.Client) error {
+	srcCluster, err := r.Plan.GetSourceCluster(c)
+	if err != nil {
+		return errors.Wrap(err, "Error reading srcMigCluster")
+	}
+
+	discovery, err := r.getDiscovery(c, srcCluster)
+	if err != nil {
+		return errors.Wrap(err, "Can't compile discovery client for srcCluster")
+	}
+
+	r.SrcDiscovery = discovery
+
+	return nil
+}
+
+func (r *GVKCompare) PrepareDestinationDiscovery(c client.Client) error {
+	dstCluster, err := r.Plan.GetDestinationCluster(c)
+	if err != nil {
+		return errors.Wrap(err, "Error reading dstMigCluster")
+	}
+
+	discovery, err := r.getDiscovery(c, dstCluster)
+	if err != nil {
+		return errors.Wrap(err, "Can't compile discovery client for dstCluster")
+	}
+
+	r.DstDiscovery = discovery
+
+	return nil
+}
+
+func (r *GVKCompare) getDiscovery(c client.Client, cluster *migapi.MigCluster) (*discovery.DiscoveryClient, error) {
+	config, err := cluster.BuildRestConfig(c)
+	if err != nil {
+		return nil, errors.Wrap(err, "Can't get REST config from a cluster")
+	}
+
+	return discovery.NewDiscoveryClientForConfig(config)
+}
+
+func (r *GVKCompare) getClient(c client.Client, cluster *migapi.MigCluster) (dynamic.Interface, error) {
+	config, err := cluster.BuildRestConfig(c)
+	if err != nil {
+		return nil, errors.Wrap(err, "Can't get REST config from a cluster")
+	}
+
+	return dynamic.NewForConfig(config)
+}
+
+func (r *GVKCompare) PrepareSourceClient(c client.Client) error {
+	srcCluster, err := r.Plan.GetSourceCluster(c)
+	if err != nil {
+		return errors.Wrap(err, "Error reading srcMigCluster")
+	}
+
+	client, err := r.getClient(c, srcCluster)
+	if err != nil {
+		return errors.Wrap(err, "Can't compile dynamic client for srcCluster")
+	}
+
+	r.SrcClient = client
+
+	return nil
+}
+
+func (r *GVKCompare) compareGroupVersions() ([]metav1.APIGroup, error) {
+	srcGroupList, err := r.SrcDiscovery.ServerGroups()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to fetch server groups for a srcCluster")
+	}
+
+	dstGroupList, err := r.DstDiscovery.ServerGroups()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to fetch server groups for a dstCluster")
+	}
+
+	missingGroups := missingGroups(srcGroupList.Groups, dstGroupList.Groups)
+	matchGroups(srcGroupList, dstGroupList, missingGroups)
+	missingVersions := missingVersions(srcGroupList.Groups, dstGroupList.Groups)
+
+	return append(missingGroups, missingVersions...), nil
+}
+
+func (r *GVKCompare) collectNamespaceReport(unsupportedResources []schema.GroupVersionResource) error {
+	for _, namespace := range r.Plan.GetSourceNamespaces() {
+		unsupportedGVRs := []string{}
+		for _, gvr := range unsupportedResources {
+			options := metav1.ListOptions{}
+			resourceList, err := r.SrcClient.Resource(gvr).Namespace(namespace).List(options)
```

Comment:
Dynamic client was the simplest option to allow generic listing of unsupported resources on the source cluster, as it does not require typecasting, and encapsulates constructing of `*Unstructured` object.